src/test/java/com/iexec/core/task/TaskServiceRealRepositoryTest.java

```diff
@@ -0,0 +1,105 @@
+import org.springframework.boot.test.autoconfigure.data.mongo.DataMongoTest;
+import org.springframework.dao.DuplicateKeyException;
+import org.springframework.test.context.DynamicPropertyRegistry;
+import org.springframework.test.context.DynamicPropertySource;
+import org.testcontainers.containers.MongoDBContainer;
+import org.testcontainers.junit.jupiter.Container;
+import org.testcontainers.junit.jupiter.Testcontainers;
+import org.testcontainers.utility.DockerImageName;
+
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.List;
+import java.util.Optional;
+import java.util.concurrent.*;
+
+import static com.iexec.core.task.TaskTestsUtils.*;
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.fail;
+
+@DataMongoTest
+@Testcontainers
+class TaskServiceRealRepositoryTest {
+    private final long maxExecutionTime = 60000;
+    private final Date contributionDeadline = new Date();
+    private final Date finalDeadline = new Date();
+
+    @Container
+    private static final MongoDBContainer mongoDBContainer = new MongoDBContainer(DockerImageName.parse("mongo:4.2"));
+
+    @DynamicPropertySource
+    static void registerProperties(DynamicPropertyRegistry registry) {
+        registry.add("spring.data.mongodb.host", mongoDBContainer::getContainerIpAddress);
+        registry.add("spring.data.mongodb.port", () -> mongoDBContainer.getMappedPort(27017));
+    }
+
+    @Autowired
+    private TaskRepository taskRepository;
+
+    @Mock
+    private IexecHubService iexecHubService;
+
+    private TaskService taskService;
+
+    @BeforeEach
+    void init() {
+        MockitoAnnotations.openMocks(this);
+        taskService = new TaskService(taskRepository, iexecHubService);
+    }
+
+    @Test
+    void shouldAddTaskASingleTime() {
+        final Task task = getStubTask(maxExecutionTime);
+        task.changeStatus(TaskStatus.INITIALIZED);
+
+        // Let's start 2 `taskService.addTask` at the same time.
+        // Without any sync mechanism, this should fail
+        // as it'll try to add twice the same task - with the same key - to the DB.
+        final ExecutorService executorService = Executors.newFixedThreadPool(2);
+        final List<Future<Optional<Task>>> executions = new ArrayList<>(2);
+        for (int i = 0; i < 2; i++) {
+            executions.add(executorService.submit(() -> taskService.addTask(CHAIN_DEAL_ID, 0, 0, DAPP_NAME, COMMAND_LINE,
+                    2, maxExecutionTime, "0x0", contributionDeadline, finalDeadline)));
+        }
+
+        // Let's wait for the `taskService.addTask` to complete.
+        executions.forEach(execution -> {
+            try {
+                execution.get(1, TimeUnit.MINUTES);
+            } catch (ExecutionException e) {
+                if (e.getCause() instanceof DuplicateKeyException) {
+                    fail("Task has been added twice. Should not happen!");
```

Comment:
I'm not sure this can happen :thinking:

Reply:
You're right. This should not happen anymore. However, I'd like to keep it there. This would ensure that, if we ever refactor `TaskService#addTask`, we still handle such a case.

Reply:
I agree. Shouldn't we try to check the `addTask` method returns the correct value as well ?

Reply:
Please see this other PR: https://github.com/iExecBlockchainComputing/iexec-core/pull/547